addons/sourcemod/scripting/freak_fortress_2.sp

```diff
@@ -6555,274 +6556,151 @@ stock GetBossIndex(client)
+	Operator_Divide,
+	Operator_Exponent,
+};
+
+stock void Operate(ArrayList sum, int bracket, float v2, ArrayList _operator)
+{
+	float oSum = sum.Get(bracket);
+	switch (_operator.Get(bracket))
+	{
+		case Operator_Add:
 		{
-			case '\0':
+			sum.Set(bracket, oSum + v2);
+		}
+		case Operator_Subtract:
+		{
+			sum.Set(bracket, oSum - v2);
+		}
+		case Operator_Multiply:
+		{
+			sum.Set(bracket, oSum * v2);
+		}
+		case Operator_Divide:
+		{
+			if (FloatCompare(v2, 0.000001) == -1 || FloatCompare(v2, -0.000001) == 1)
 			{
-				foundOperator = true; //Probably less hacky way of detecting constant hp.
-				Debug("Null terminator found!");
+				LogError("[FF2] Cowardly refusing divide by zero!");
+				return;
 			}
+			sum.Set(bracket, oSum / v2);
+		}
+		case Operator_Exponent:
+		{
+			sum.Set(bracket, Pow(oSum, v2));
+		}
+		default:
+		{
+			sum.Set(bracket, v2);
+		}
+	}
+	_operator.Set(bracket, Operator_None);
+}
+
+stock void OperateString(ArrayList sum, int bracket, char[] strValue, int size, ArrayList _operator)
+{
+	if (StrEqual(strValue, ""))
+	{
+		return;
+	}
+	Operate(sum, bracket, StringToFloat(strValue), _operator);
+	strcopy(strValue, size, "");
+}
+
+stock float IterateFormula(const char[] formula, const char[] key, const char[] bossName)
+{
+	ArrayList sum = new ArrayList();
+	ArrayList _operator = new ArrayList();
+	int bracket;
+	char buffer[2], strValue[1024];
+	for (int i = 0; i <= strlen(formula); i++)
+	{
+		buffer[0] = formula[i];
+		switch (buffer[0])
+		{
 			case '(':
 			{
-				Debug("'(' found! bracket = %i", bracket);
-				if (bracket + 1 == sizeof(_operator))
+				++bracket;
+				if (sum.Length < bracket)
 				{
-					decl String:bossName[32];
-					KvRewind(BossKV[Special[client]]);
-					KvGetString(BossKV[Special[client]], "name", bossName, sizeof(bossName));
-					LogError("[FF2] %s's %s formula is too complex, using default!", bossName, key);
-					return defaultValue;
+					sum.Resize(bracket);
+					_operator.Resize(bracket);
 				}
-				closeBracket = false;
-				bracket++;
-				_operator[bracket] = Operation_None;
-				sum[bracket] = 0.0;
+				sum.Set(bracket, 0.0);
+				_operator.Set(bracket, Operator_None);
 			}
 			case ')':
 			{
-				Debug("')' found!");
-				closeBracket = true;
-			}
-			case 'n', 'x':
-			{
-				Debug("We must operate with players!");
-				usePlayers = true;
-			}
-			case '+':
-			{
-				Debug("Addition time!");
-				_operator[bracket] = Operation_Add;
-				foundOperator = true;
-			}
-			case '-':
-			{
-				Debug("Subtraction time!");
-				_operator[bracket] = Operation_Subtract;
-				foundOperator = true;
-			}
-			case '*':
-			{
-				Debug("Multiplication time!");
-				_operator[bracket] = Operation_Multiplication;
-				foundOperator = true;
-			}
-			case '/':
-			{
-				Debug("Division time!");
-				_operator[bracket] = Operation_Division;
-				foundOperator = true;
+				OperateString(sum, bracket, strValue, sizeof(strValue), _operator);
+				if ((--bracket) < 0)
```

Comment:
It has to do with the difference between ++ infront of var and ++ after
On 2015-03-21 9:17 PM, "Wliu" <notifications@github.com> wrote:

> In addons/sourcemod/scripting/freak_fortress_2.sp
> <https://github.com/50DKP/FF2-Official/pull/232#discussion_r26898776>:
>
> > -				_operator[bracket] = Operation_Subtract;
> > -				foundOperator = true;
> > -			}
> > -			case '*':
> > -			{
> > -				Debug("Multiplication time!");
> > -				_operator[bracket] = Operation_Multiplication;
> > -				foundOperator = true;
> > -			}
> > -			case '/':
> > -			{
> > -				Debug("Division time!");
> > -				_operator[bracket] = Operation_Division;
> > -				foundOperator = true;
> > +				OperateString(sum, bracket, strValue, sizeof(strValue), _operator);
> > +				if ((--bracket) < 0)
>
> Any reason this can't be bracket--<0
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/50DKP/FF2-Official/pull/232/files#r26898776>.
>